Add equal_to_1 to triton_meta for user-written Triton kernels

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120579

Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `@user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames

aakhundov committed:
Add equal_to_1 to triton_meta for user-written Triton kernels

Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `@user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478.

Before

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

[ghstack-poisoned]

oulgen commented on test/dynamo/test_triton_kernels.py:
```diff
@@ -958,6 +959,21 @@ def f(x):
         compiled_out = torch.compile(f, fullgraph=True, backend=backend)(x)
         self.assertEqual(compiled_out, eager_out)
 
+    def test_triton_kernel_equal_to_1_arg(self):
```
can you add a test with dynamic shapes too?

aakhundov replied:
Extended the test to cover dynamic shapes.

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov commented:
Fixed broken tests; added another new test for AOT Inductor (to make sure setting `equal_to_1` and constants doesn't break anything in AOT compilation and C++ codegen).

@oulgen I've excluded adding `None` arguments to constants from the scope of this PR, as it causes CUDA errors in the `test_aot_inductor.test_triton_kernel_with_none_input`. This will require a separate investigation, so I've left a TODO in this PR.

jansel commented on torch/_inductor/codegen/triton_utils.py:
```diff
@@ -63,7 +63,12 @@ def signature_to_meta(
+    args: List[KernelArgType],
+    *,
+    indices: Optional[List[int]] = None,
+    add_equal_to_1: bool = False,
```
Why not just add this always?

aakhundov replied:
My initial assumption was that, if this had not been handled before for the Inductor-generated Triton kernels, then there must have been some issue there. Let me try adding always and see how the CI responds.

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov commented:
@jansel I've tried enabling this for Inductor-generated kernels, too, and some tests started failing (mostly due to changes in the codegen: like two kernels generated instead of one). I've patched them, and now there are more failing tests. I can continue patching those, but my feeling is that for Inductor-generated kernels the change can lead to unexpected regressions. Would it make sense to limit the change to user-written Triton kernels in this PR and extend to Inductor-generated ones in a follow-up?

jansel commented:
@aakhundov I want to make sure this change is correct and we aren't just hiding bugs by disabling it most of the time.  It is a bit worrying it is leading to other failures.  We already 1-specialize, so there shouldn't be any extra kernels.

aakhundov commented:
@jansel makes sense. By "extra kernels" I mean, due to the `equal_to_1` being populated for the call sites where 1 is passed to the kernel, now the overall source of the kernel (including the `triton_meta` part in the decorator) will change for those specific call sites. As a result, [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L3497) we generate a new kernel in the codegen with the same source, except the `equal_to_1` specialization. This is the reason of the failing tests I've patched. But now I also see some tests failing the accuracy checks. I'll need to investigate and fix it before we can land this PR for all kernels, both user-written and Inductor-generated.

jansel commented:
Sounds good.  We should look into that accuracy issue.  If it turns out to be a Triton bug we can pull in some other folks to help.

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov commented:
@jansel Further investigation has shown that, when arguments are included into `equal_to_1` and `constants`, they are skipped in the PTX. E.g., for the same kernel with empty `equal_to_1` in `triton_meta`:

```
@reduction(
    size_hints=[1, 1024],
    reduction_hint=ReductionHint.INNER,
    filename=__file__,
    triton_meta={'signature': {0: '*fp32', 1: '*i64', 2: 'i32', 3: 'i32'}, 'device': 0, 'device_type': 'cuda', 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1), equal_to_1=(), ids_of_folded_args=(), divisible_by_8=())]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_red_fused_ne_sum_0', 'mutated_arg_names': [], 'no_x_dim': False}
)
```

we get four parameters in the PTX:

```
.visible .entry triton__0d1d23(
	.param .u64 triton__0d1d23_param_0,
	.param .u64 triton__0d1d23_param_1,
	.param .u32 triton__0d1d23_param_2,
	.param .u32 triton__0d1d23_param_3
)
```

but when we include one of the arguments into `equal_to_1` and `constants`:

```
@reduction(
    size_hints=[1, 1024],
    reduction_hint=ReductionHint.INNER,
    filename=__file__,
    triton_meta={'signature': {0: '*fp32', 1: '*i64', 2: 'i32', 3: 'i32'}, 'device': 0, 'device_type': 'cuda', 'constants': {2: 1}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1), equal_to_1=(2,), ids_of_folded_args=(2,), divisible_by_8=())]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_red_fused_ne_sum_0', 'mutated_arg_names': [], 'no_x_dim': False}
)
```

we see just three parameters in the PTX:

```
.visible .entry triton__0d1d2c3(
	.param .u64 triton__0d1d2c3_param_0,
	.param .u64 triton__0d1d2c3_param_1,
	.param .u32 triton__0d1d2c3_param_2
)
```

This was the root cause of the accuracy issues in the AOT Inductor / C++ wrapper codegen tests. I've added skipping 1 arguments when launching AOT-compiled Triton kernels in the C++ wrapper codegen. This seems to have resolved the CI issues.

Could you take another look at the PR? Thanks!

chenyang78 commented on torch/_inductor/codegen/cpp_wrapper_cuda.py:
```diff
@@ -264,6 +264,11 @@ def generate_kernel_call(
+        # args with value 1 are added into equal_to_1 and constants
+        # in triton_meta (in the Python codegen) which makes them
+        # inlined in the PTX and compiled CUBIN
+        call_args = [arg for arg in call_args if arg != 1]
```
not a blocker, but just a random thought - instead of skipping ```arg == 1``` cases, would it possible to not generate those args whose value equals ```1```? My worry is that it could potentially introduce mismatch (e.g.) if the same call_args is used somewhere else but we didn't filter out ```arg == 1``` arguments.

Second, I assume that all ```arg == 1``` will be added into ```equal_to_1``` meta? Thanks.

jansel replied:
Agreed it would be better to use the hints to guarantee there isn't a mismatch.

aakhundov replied:
> Second, I assume that all arg == 1 will be added into equal_to_1 meta?

This is correct. That's why we are excluding them here with certainty.

> would it possible to not generate those args whose value equals 1? 

I don't think we can skip those args in the upstream generation (before the codegen), as the 1-args must still be passed in the Python wrapper, but skipped in the C++ wrapper (due to the PTX nuance mentioned above). Could you point to other places in the C++ codegen where we might still need to exclude 1-args? Thanks!

chenyang78 replied:
> > Second, I assume that all arg == 1 will be added into equal_to_1 meta?
> 
> This is correct. That's why we are excluding them here with certainty.
> 
> > would it possible to not generate those args whose value equals 1?
> 
> I don't think we can skip those args in the upstream generation (before the codegen), as the 1-args must still be passed in the Python wrapper, but skipped in the C++ wrapper (due to the PTX nuance mentioned above). Could you point to other places in the C++ codegen where we might still need to exclude 1-args? Thanks!

Thanks for the clarification. I am not aware of any other use of ```call_args``` in the current cuda wrapper code. It's just my gut-feeling that dropping args based on unguarded assumptions seems to be dangerous :)

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov commented:
@jansel @chenyang78 I've changed the codegen part to skip args explicitly based on the contents of `equal_to_1` hint propagated to the codegen within the `triton_meta`. Please take a look. Thanks!

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

chenyang78 reviewed (approved):
LGTM. Thanks!

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov committed:
Update on "Add equal_to_1 to triton_meta for user-written Triton kernels"


Summary: Previously, we omitted `equal_to_1` from the `triton_meta` part of the `user_autotune` decorator. For user-written Triton kernels, this could lead to perf regressions, as the kernel in the Inductor codegen is compiled without `equal_to_1` specialization.

Fixes #120478. The repro from the issue, on A100:

Before this PR:

```
Triton matmul:           0.0167 seconds
Triton matmul compiled:  0.0751 seconds
```

After this PR:

```
Triton matmul:           0.0168 seconds
Triton matmul compiled:  0.0072 seconds
```

Test Plan:

```
$ python test/dynamo/test_triton_kernels.py -k  test_triton_kernel_equal_to_1_arg
...
----------------------------------------------------------------------
Ran 3 tests in 3.545s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge

aakhundov commented:
@pytorchbot merge -f "unrelated failures"
